qiita_db/processing_job.py

```diff
@@ -6,21 +6,295 @@
+            value = search('<%s>(.*?)</%s>' % (element, element), snippet)
+            if value:
+                results[element] = value.group(1)
+            else:
+                raise AssertionError("%s element is missing!" % element)
+
+        for element in list_of_optional_elements:
+            value = search('<%s>(.*?)</%s>' % (element, element), snippet)
+            if value:
+                results[element] = value.group(1)
+
+        return results
+
+    def _process_dependent_jobs(self, results):
+        # when a job has its status changed, check to see if the job completed
+        # with an error. If so, check to see if had any jobs that were being
+        # 'held' on this job's successful completion. If we are maintaining
+        # state on any of these jobs, mark them as 'DROPPED', because they will
+        # no longer appear in qstat output.
+        if results['job_state'] == 'completed':
+            if results['exit_status'] == '0':
+                return
+
+            if 'depend' in results:
+                tmp = results['depend'].split(':')
+                if tmp[0] == 'beforeok':
+                    tmp.pop(0)
+                    for child_job_id in tmp:
+                        # jobs in 'beforeok' are labeled with the complete
+                        # job id and what looks to be the server name doing
+                        # the work. For now, simply remove the
+                        # '@host.domain.org' (server) component.
+                        child_job_id = child_job_id.split('@')[0]
+                        self.processes[child_job_id]['job_state'] = 'DROPPED'
+                        self.queue.put(self.processes[child_job_id])
+
+    def run(self):
+        # check to see if qstat is available. If not, exit immediately.
+        proc = Popen("qstat -x", shell=True, stdout=PIPE, stderr=PIPE)
+        stdout, stderr = proc.communicate()
+        if proc.returncode != 0:
+            # inform any process expecting data from Watcher
+            self.queue.put('QUIT')
+            self.event.set()
+
+        while not self.event.is_set():
+            proc = Popen("qstat -x", shell=True, stdout=PIPE, stderr=PIPE)
+            stdout, stderr = proc.communicate()
+            if proc.returncode == 0:
+                # qstat returned successfully with metadata on processes
+                # break up metadata into individual <Job></Job> elements
+                # for processing.
+                m = findall('<Job>(.*?)</Job>', stdout)
+                for item in m:
+                    # filter out jobs that don't belong to owner
+                    if search('<Job_Owner>%s</Job_Owner>' % self.owner, item):
+                        # extract the metadata we want.
+                        # if a job has completed, an exit_status element will
+                        # be present. We also want that.
+                        results = self._element_extract(item, ['Job_Id',
```

Comment:
Did you explore parsing the object with Python's XML parser? Maybe that would make some of these tasks more straightforward? Alternatively maybe parsing the output of qstat as a pandas dataframe?

Reply:
I considered it, but it did seem relatively straightforward at the time to just extract the few lines needed based on regex instead. Scraping the data into a pandas df wasn't desirable, as the XML output provides extended information that the tabular presentation doesn't provide, at least by default.